### ISSUE TITLE
Metadata handling through processes

### ISSUE THREAD

**Issue by huard** [cf-convention/discuss#33]
We're building processes operating on netCDF files, and wondering what are the best-practices in terms of metadata handling. 

Imagine the following case: 
```
input 1 -> process A \
                       process B -> output
input 2 -> process A / 
```

Let's say for example that `input 1` and 2  have CMIP-like metadata. What should happen to this metadata as it goes through processes 1 and 2? I suppose that different attributes would go through either: 
- Aggregation
- Updates
- Replacement
- Deletion 

For CMIP datasets we could save the PIDs from the original files, but if an input is observed data, it might not have a PID or DOI. Anyway, curious to know how to approach this systematically so that it scales to large ensembles and complex workflows with multiple processes.

**Comment by huard** [comment]
One solution to this issue would be for CF to adopt a "provenance" standard (e.g. PROV) and store a machine-readable provenance string in global or variable attributes.

**Comment by JonathanGregory** [comment]
CF has attributes which could be used for recording that kind of history, but their contents are not standardised, such as `source`, `history` and `comment`. They are in section 2.6.2 of the standard.

**Comment by huard** [comment]
Yes, but they are meant to be human-readable. I think using them to store XML would abuse their intended meaning. Also, I think it would cause all kinds of problems for frontends that would have to check whether these attributes are human or machine readable before displaying them.

**Comment by sfoucher** [comment]
Maybe the new OGCApi Record could be useful here (https://github.com/opengeospatial/ogcapi-records) at least for observations such as satellite imagery. In the metadata of B you could point to records from inputs 1 and 2 for instance.

**Comment by cofinoa** [comment]
@huard additionally to human readable attributes, the `cell_methods` variable attribute also provides a machine friendly process (i.e. aggregation) processes. In fact all section 7: 
http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#_data_representative_of_cells
define metadata about processes applied to data. 
IMO, the "provenance" proposal should be consider those as an alternative or supplementary metadata about processes been applied to data. 

BTW, @larsbarring it's also working to encode metadata about process been applied to data, and in particular, computation on climate indexes. It may be would be useful to have his opinion on this.

**Comment by davidhassell** [comment]
Just to highlighting the synergy that Lars's issue on climate indices (https://github.com/cf-convention/cf-conventions/issues/197) is also going to be discussed at the CF meeting next week, as well as this one.

**Comment by huard** [comment]
Agreed, this is not meant to replace existing attributes, but as you suggest, as supplementary information about processes. 

To clarify, a use case could be to describe operations leading to the creation of a map of forest fire under future climate conditions. The provenance would list the source data from say 20 climate model projections, a regridding algorithm used to interpolate them on a common grid, a bias-adjustment algorithm and an observation dataset used to remove model biases, details about the fire model used to process the meteorological input, and then statistics applied to the ensemble of fire projections.

Although it's possible to capture the details of all these processes into a "history" field, my experience is that it becomes unreadable. So I see "history" as being a high-level human readable synthesis of operations, and provenance as the low-level machine-readable full description of operations and source files.

**Comment by larsbarring** [comment]
I think that the connection between the two issues is in the second bulleted point of the breakout intro:

"...the more far-reaching need for a new or alternative mechanism that allows for a more flexible description of more complex and/or multi-step temporal processing of data."

But I also think that it would be difficult to capture all the relevant aspects that you mention within a framework that could be thought of as an extension of the CF Conventions. If I understand what you are aiming for, it is a far-reaching and open-ended need.

**Comment by davidhassell** [comment]
Hello,

The timings and order of the breakout groups for the CF meeting next week has now been set (see http://cfconventions.org/Meetings/2020-Workshop.html), and the discussion of this issue will be on Thursday 11 June from 16:00-17:30 UTC, in parallel with three other topics.

Thanks.

**Comment by ngalbraith** [comment]
Regarding a new 'provenance' attribute, I'd like to suggest that we might consider allowing this metadata to be stored as a variable, instead of an attribute. 

This  gives the data producer more ability, when combining datasets along geographical or depth axes, to easily keep the provenance metadata with the correct part of the data, using depth or x/y indexes - hard to do this in an attribute. E.g. the salinity below a certain depth might have pressure corrections applied differently from other depths ... documenting that in a 2D salinity file using variable attributes (especially in a machine-readable form) is pretty much impossible to do through automatic processes.

Also, sometimes it's appropriate to provide provenance as a data variable attribute, and sometimes it's more appropriate as a global, but in a situation where a single term appears as both a global and a variable attribute, most software does not behave as expected (according to an old discussion on the CF list).

**Comment by huard** [comment]
Edited the proposal to include this idea.

**Comment by zklaus** [comment]
Another effort that might be related is [OGC netCDF-LD](https://github.com/opengeospatial/netCDF-LD/), a draft standard to encode RDF in netCDF. This could potentially be used to store [PROV-O](https://www.w3.org/TR/prov-o/) information directly in netCDF.

**Comment by huard** [comment]
Thanks @zklaus, 

I agree it would work, but I'm somewhat skeptical about its adoption by scientists. The syntax is not easy on the eye and I think all these "bald__" attributes would cause confusion for users. So although I think we should consider it in this discussion, I'd like to propose something that does not even try to be human-readable.

**Comment by zklaus** [comment]
@huard, then XML is definitely your go-to ;)

Yeah, I only meant to bring it up for consideration, particularly since OGC seems to be something that comes up in several places here and since it seems that @ethanrd is involved (see [OGC netCDF SWG](https://www.ogc.org/projects/groups/netcdfswg)).

**Comment by huard** [comment]
Totally agree that whatever is chosen should be OGC-compliant.

**Comment by jbedia** [comment]
Hi everyone, thanks for this interesting discussion.
Following the meeting held yesterday, this is just a small sample of a RDF - JSON file containing the provenance information of a CMIP5 ensemble map of a bias-adjusted climate index based on maximum temperature. For simplicity, it is formed by only two GCMs. https://drive.google.com/file/d/1ij-JKNF2lFidbM823jwDQcOxDz4Ntx4m/view?usp=sharing

It can for instance be dropped onto the [www.metaclip.org](www.metaclip.org) interpreter to be explored, but its complexity, with just two models, is huge. With an ensemble of ~30 models it is not possible to go for the full detail provenance information visually, and somehow the information needs to be collapsed/summarized. We are currently working on that part.
An advantage of such RDF provenance representation is that it allows to expose the provenance information at different levels of granularity. It also ensures interoperability. The RDF graph can be serialized onto many different formats, not just JSON, but also for instance XML.
There are some examples of climate products available in [www.metaclip.org](www.metaclip.org) with attached provenance information that are automatically displayed in the viewer. However, the provenance information can be also linked externally via a reference attribute, as recommended for netCDF/NcML files.
Comments and suggestions are very welcome

**Comment by huard** [comment]
Trying to push the idea of looking into provenance in OGC Testbed 17: https://github.com/opengeospatial/ideas/issues/113
